src/types/src/parser.c

```diff
@@ -0,0 +1,152 @@
+#include "os/os.h"
+
+#define YYSTYPE DDS_TS_PARSER_STYPE
+#define YYLTYPE DDS_TS_PARSER_LTYPE
+
+#define YY_TYPEDEF_YY_SCANNER_T
+typedef void* yyscan_t;
+
+#define YY_NO_UNISTD_H 1 /* to surpress #include <unistd.h> */
+
+#include "parser.h"
+#include "idl.parser.h"
+#include "yy_decl.h"
+#include "idl.lexer.h"
+
+extern void dds_ts_stringify(dds_ts_node_t *root_node, char *buffer, size_t len);
+
+extern int
+dds_ts_parse_file(const char *file, void (*error_func)(int line, int column, const char *msg))
+{
+  int err = 0;
+  FILE *fh;
+
+  assert(file != NULL);
+
+OS_WARNING_MSVC_OFF(4996);
+  fh = fopen(file, "rb");
+OS_WARNING_MSVC_ON(4996);
+
+  if (fh == NULL) {
+    err = errno;
+    if (error_func != 0)
+      error_func(0, 0, "Cannot open file");
+  } else {
+    dds_ts_context_t *context = dds_ts_create_context();
+    if (context == NULL) {
+      if (error_func != 0)
+        error_func(0, 0, "Error: out of memory\n");
+      return 2;
+    }
+    dds_ts_context_set_error_func(context, error_func);
+    yyscan_t scanner;
+    dds_ts_parser_lex_init(&scanner);
+    dds_ts_parser_set_in(fh, scanner);
+    err = dds_ts_parser_parse(scanner, context);
+    if (err == 0) {
+      char buffer[1000];
+      dds_ts_stringify(dds_ts_context_get_root_node(context), buffer, 1000);
+OS_WARNING_MSVC_OFF(4996);
+      printf("Result: '%s'\n", buffer);
+OS_WARNING_MSVC_ON(4996);
+    }
+    else if (dds_ts_context_get_out_of_memory_error(context)) {
+      if (error_func != 0)
+        error_func(0, 0, "Error: out of memory\n");
+    }
+    dds_ts_free_context(context);
+    dds_ts_parser_lex_destroy(scanner);
+    (void)fclose(fh);
+  }
+
+  return err;
+}
+
+extern int
+dds_ts_parse_string(const char *str, void (*error_func)(int line, int column, const char *text))
+{
+  int err = 0;
+
+  assert(str != NULL);
+  if (str == NULL) {
+    if (error_func != NULL)
+      error_func(0, 0, "String argument is NULL");
+    err = -1;
+  } else {
+    dds_ts_context_t *context = dds_ts_create_context();
+    if (context == NULL) {
+      if (error_func != NULL)
+        error_func(0, 0, "Out of memory");
+      return 2;
+    }
+    dds_ts_context_set_error_func(context, error_func);
+    yyscan_t scanner;
+    dds_ts_parser_lex_init(&scanner);
+    dds_ts_parser__scan_string(str, scanner);
+    err = dds_ts_parser_parse(scanner, context);
+    if (err != 0) {
+      if (dds_ts_context_get_out_of_memory_error(context))
+        if (error_func != NULL)
+          error_func(0, 0, "Out of memory");
+    }
+    dds_ts_free_context(context);
+    dds_ts_parser_lex_destroy(scanner);
+  }
+
+  return err;
+}
+
+/* For testing: */
+
+int dds_ts_parse_string_stringify(const char *str, char *buffer, size_t len)
+{
+  int err = 0;
+
+  assert(str != NULL);
+  assert(buffer != NULL);
+
+  if (str == NULL || buffer == NULL) {
+    err = -1;
+  } else {
+    dds_ts_context_t *context = dds_ts_create_context();
+    if (context == NULL) {
+      if (len > 20)
```

Comment:
@FransFaase, this is a rather funny statement in my opinion. strlcpy always nul-terminates and the `len > 20` is chosen arbitrarily unless I'm missing something. Please assert len is larger than zero (better to check if it's large enough to hold the error message) before doing anything and remove this statement. Also, not a fan of the magic values being used to indicate errors. I'm ok with it for now because it's merely for testing and I'm working on exporting error codes from the abstraction layer.

Reply:
Indeed, that statement should have been removed.
With respect to the error codes returned, I will wait for your changes. I expect that some of the parsing functions will move to the tool in the future.